content/post/2020-02-07-processx-blocked-sigchld-fix/index.md

```diff
@@ -0,0 +1,458 @@
+`SIGCHLD` signal:
+
+```
+(lldb) call (void*) malloc(sizeof(sigset_t))
+(void *) $3 = 0x00007f92704c4300
+(lldb) call (int) sigprocmask(0, NULL, $3)
+(int) $4 = 0
+(lldb) p (int) sigismember((const sigset_t*) $3, 20)
+(int) $7 = 1
+```
+
+`SIGCHLD` is indeed blocked! We got it!
+
+## Fixing the bug
+
+Clearly, if `base::system()` works and `callr::r()` and `processx::run()`
+do not, that means that processx causes the `SIGCHLD` to be blocked in the
+subprocess.
+
+When processx starts a subprocess, it first calls `fork(2)` to create a
+copy of the current process, and then `execvp(3)` to replace that with
+another executable. `fork()` creates an identical copy, i.e. the signal
+handlers and the blocked signals are the same in the subprocess.
+`execvp(3)` resets the signal handlers to their defaults, which is great,
+but apparently, it does not reset the set of blocked signals. So if
+`SIGCHLD` was blocked when processx called `fork(2)` then it will be blocked
+in the subprocess as well. Indeed,
+[this is how processx started the subprocess](https://github.com/r-lib/processx/blob/8baaa7ead58a39fda72974ccb0e5a288d9757e4a/src/unix/processx.c#L486-L511)
+before fixing this bug:
+
+```c
+  processx__block_sigchld();
+
+  pid = fork();
+
+  if (pid == -1) {		/* ERROR */
+    err = -errno;
+    if (signal_pipe[0] >= 0) close(signal_pipe[0]);
+    if (signal_pipe[1] >= 0) close(signal_pipe[1]);
+    if (cpty) close(pty_master_fd);
+    processx__unblock_sigchld();
+    R_THROW_SYSTEM_ERROR_CODE(err, "Cannot fork when running '%s'",
+                              ccommand);
+  }
+
+  /* CHILD */
+  if (pid == 0) {
+    if (cpty) close(pty_master_fd);
+    processx__child_init(handle, pipes, num_connections, ccommand, cargs,
+                         signal_pipe[1], cstdin, cstdout, cstderr,
+                         pty_name, cenv, &options, ctree_id);
+    R_THROW_SYSTEM_ERROR("Cannot start child process when running '%s'",
+                         ccommand);
+  }
+
+```
+
+processx blocks `SIGCHLD` before forking, for simplicity, so it does not
+have to worry about concurrency. `processx__child_init()` does a number
+of things, but it does _not_ unblock this signal, so it stays blocked.
+
+The [fix](https://github.com/r-lib/processx/commit/d5369fdd0075ff9570b2ea0405fa897624a91b94)
+is clearly to unblock the signal in the child process, before calling
+`execvp(3)` in `processx__child_init()`:
+
+```c
+  /* CHILD */
+  if (pid == 0) {
+    /* LCOV_EXCL_START */
+    if (cpty) close(pty_master_fd);
+    processx__unblock_sigchld();
+    processx__child_init(handle, pipes, num_connections, ccommand, cargs,
+			 signal_pipe[1], cstdin, cstdout, cstderr,
+                         pty_name, cenv, &options, ctree_id);
+    R_THROW_SYSTEM_ERROR("Cannot start child process when running '%s'",
+                         ccommand);
+    /* LCOV_EXCL_STOP */
+  }
+```
+
+## Conclusion
+
+Debugging is hard, but debuggers like lldb help you greatly. It is worth
```

Comment:
Could you add one or more resources to learn that?

Reply:
I know of https://r-pkgs.org/src.html#src-debugging

Reply:
OK, I'll add some links.